promotion.md

```diff
@@ -111,6 +121,41 @@ it is unlikely to be the desired behavior in most cases and very likey to be con
 to the user. If such behavior is desired, the user can still use an explicit `static`
 or `const` item and refer to that.
 
+*Dynamic check.* The Miri engine could dynamically check this by ensuring that
+ the result of computing a promoted is a value that does not need dropping.
+
+## `&` in `const` and `static`
+
+Promotion is also responsible for making code like this work:
+
+```rust
+const FOO: &'static i32 = {
+    let x = &13;
+    x
+};
+```
+
+However, since this is in explicit const context, we could be less strict about
```

Comment:
We are less strict. You can promote any const fn call inside a constant

Reply:
Ah, good to know! Are there any other differences? We could in principle be even more aggressive.

Reply:
iirc we are running just the regular const checks on promoteds inside constants and const fns. So there's no third system, it's exactly as if you wrote an explicit constant

Reply:
Isnt promotion also about identifying which parts of the MIR to promote, and then actually patch the MIR to replace uses of the data by a promoted? Somehow this still needs to do something extra inside constants I think, or will it just trigger at every `&`?

Reply:
So implementation wise the difference is `IsPromotable` vs `IsImplicitlyPromotable`, where iirc `IsImplicitlyPromotable` only makes a difference in non-const fns

Reply:
So `IsPromotable` is *not* about the kind of promotion I am talking about here? :(
What is it about then?

I asked @eddyb about this terminology but must have misunderstood what he said.

Reply:
> So IsPromotable is not about the kind of promotion I am talking about here? :(

I'm confused.

So `IsPromotable` and `IsImplicitlyPromotable` are required in normal functions, while `const fn` and constants only check `IsPromotable`

Reply:
I guess I am confused by the "implicit" part. The promotion to a separate static is also implicit inside a const body.

If the thing inside a normal `fn` is "implicit promotion", I would expect the thing inside a const body to be "explicit promotion", but that does not make sense, does it?

Reply:
yea... it's not the best naming scheme... Maybe we should call the thing in non-const fns "restricted promotion" and the other cases just "promotion" or "unrestricted promotion"

Reply:
That doesn't convey why it is restricted, though. "implicit" did that nicely, but was not precise enough.

"promotion of run-time code" is verbose but informative. "run-time promotion" is somewhat misleading... "run-time code promotion"?